cognite/powerops/resync/v2/shop_to_assets.py

```diff
@@ -0,0 +1,370 @@
+            if reservoir.external_id in existing:
+                raise ValueError(f"Reservoir with external id {reservoir.external_id} already exists")
+            reservoir_by_name[reservoir.name] = reservoir
+            assets_by_xid[reservoir.external_id] = reservoir
+
+        plants = []
+        for name, data in shop_model["model"]["plant"].items():
+            plant = self._to_plant(
+                name,
+                data,
+                plant_display_name_and_order,
+                watercourse_external_id,
+                reservoir_by_name,
+                generator_by_name,
+                shop_model,
+            )
+            if plant.external_id in existing:
+                raise ValueError(f"Plant with external id {plant.external_id} already exists")
+            plants.append(plant)
+            assets_by_xid[plant.external_id] = plant
+
+        data = watercourse_config
+        watercourse = WatercourseWrite(
+            external_id=watercourse_external_id,
+            name=data["name"],
+            production_obligation=data.get("production_obligation_ts_ext_ids", []),
+            penalty_limit=data.get("shop_penalty_limit", self.default_shop_penalty_limit),
+            plants=plants,
+        )
+        if watercourse.external_id in existing:
+            raise ValueError(f"Watercourse with external id {watercourse.external_id} already exists")
+        assets_by_xid[watercourse.external_id] = watercourse
+
+        for price_area in watercourse_config.get("market_to_price_area", {}).values():
+            if price_area in existing:
+                raise ValueError(f"Price area with external id {price_area} already exists")
+            price_area = PriceAreaWrite(
+                external_id=f"price_area_{price_area}", name=price_area, timezone=self.default_timezone
+            )
+            assets_by_xid[price_area.external_id] = price_area
+
+        return assets_by_xid
+
+    def _to_generator(self, name: str, data: dict) -> GeneratorWrite:
+        curve = data["gen_eff_curve"]
+        efficiency_curve = GeneratorEfficiencyCurveWrite(
+            external_id=f"{name}_efficiency_curve", ref=curve["ref"], power=curve["x"], efficiency=curve["y"]
+        )
+        turbine_curves = []
+        for curve in data["turb_eff_curves"]:
+            turbine_curve = TurbineEfficiencyCurveWrite(
+                external_id=f"{name}_turbine_eff_curve_ref_{curve['ref']}",
+                head=curve["ref"],
+                flow=curve["x"],
+                efficiency=curve["y"],
+            )
+            turbine_curves.append(turbine_curve)
+
+        # Start cose is assumed to be a dictionary with (timestamp, value) key-value pairs
+        # We assume it is constant and thus use only the first value.
+        startcost = next(iter(data["startcost"].values()))
+
+        generator_timeseries = self.times_series_by_generator_name.get(name, {})
```

Comment:
**issue:** should throw an error because an empty dict is not valid

Reply:
line 175 & 176 will throw a KeyError if an empty dict is returned

Reply:
sorry will return None but again not clear here that that's intended behavior

Reply:
It is.